src/mpid/ch4/shm/ipc/src/ipc_send.h

```diff
@@ -23,31 +23,65 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_IPCI_try_lmt_isend(const void *buf, MPI_Aint
 
     MPID_THREAD_CS_ENTER(VCI, MPIDI_VCI(0).lock);
 
+    if (rank == comm->rank) {
+        goto fn_exit;
+    }
+
+    MPIR_Datatype *dt_ptr;
     bool dt_contig;
     MPI_Aint true_lb;
     uintptr_t data_sz;
-    MPIDI_Datatype_check_contig_size_lb(datatype, count, dt_contig, data_sz, true_lb);
+    MPIDI_Datatype_get_info(count, datatype, dt_contig, data_sz, dt_ptr, true_lb);
+
+    if (!dt_contig && (dt_ptr->true_lb < 0 || dt_ptr->extent < 0)) {
+        /* TODO: handle offset */
+        goto fn_exit;
+    }
 
-    void *vaddr = (char *) buf + true_lb;
+    MPI_Aint mem_size;
+    void *mem_addr;
+    if (dt_contig) {
+        mem_addr = (char *) buf + true_lb;
+        mem_size = data_sz;
+    } else {
+        mem_addr = (char *) buf;
+        mem_size = dt_ptr->true_ub + (count - 1) * dt_ptr->extent;
+    }
     MPIDI_IPCI_ipc_attr_t ipc_attr;
-    MPIR_GPU_query_pointer_attr(vaddr, &ipc_attr.gpu_attr);
+    MPIR_GPU_query_pointer_attr(mem_addr, &ipc_attr.gpu_attr);
 
     if (ipc_attr.gpu_attr.type == MPL_GPU_POINTER_DEV) {
-        mpi_errno = MPIDI_GPU_get_ipc_attr(vaddr, rank, comm, &ipc_attr);
+        mpi_errno = MPIDI_GPU_get_ipc_attr(mem_addr, rank, comm, &ipc_attr);
```

Comment:
FIXME: With non-contig datatypes, this check is insufficient.  We need at least quiry two address, one start address and one end address. We can only do single map IPC when the entire address range falls within the same device.

In theory, when different segments are from different device, we still can do multiple IPC -- one for each device. But that will complicate the code significantly and may incur significant query cost when the number of segments are huge. It may also be a non-common case anyway. Thus we probably should just skip this optimization.